File: py_drill/Problems/TR08_BinaryTreeMaxPathSum.py
# REF : https://leetcode.com/problems/binary-tree-maximum-path-sum/
# NOTES :
#       Run DFS & while rolling back calculate max sum so far and store it in a max_sum variable.
#       max_sum will be -> max including this current node OR excluding this current node
#       (i.e. max_sum so far) -> max(maxSum, node.val + left_max + right_max)
#       while returning, it will return including current node val + max so far
#       (i.e. max(left_max,right_max)

# T: O(n)
# S: O(H) for balanced tree OR O(n) for skew tree
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root) -> int:

        def dfs(node):
            nonlocal maxSum
            if not node:
                return 0

            left_max = dfs(node.left)
            right_max = dfs(node.right)

            left_max = max(left_max, 0)  # This will avoid -ve or sum reducing cases
            right_max = max(right_max, 0)

            maxSum = max(maxSum, node.val + left_max + right_max)

            return node.val + max(left_max, right_max)

        maxSum = float("-inf")
        dfs(root)
        return maxSum


def main():
    pass

if __name__ == "__main__":
    main()

File: cpp_drill/Problems/TR08_BinaryTreeMaxPathSum.cpp
#include <iostream>
using namespace std;

// REF : https://leetcode.com/problems/binary-tree-maximum-path-sum/
// NOTES :
//       Run DFS & while rolling back calculate max sum so far and store it in a max_sum variable.
//       max_sum will be -> max including this current node OR excluding this current node
//       (i.e. max_sum so far) -> max(maxSum, node.val + left_max + right_max)
//       while returning, it will return including current node val + max so far
//       (i.e. max(left_max,right_max)
//       Concept similar to Diameter of Binary Tree.

// T: O(n)
// S: O(n)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int max_sum = INT_MIN;
        dfs(root, max_sum);
        return max_sum;
    }

    int dfs(TreeNode* node, int& max_sum) {
        if (node == nullptr) return 0;

        int left_max = dfs(node->left, max_sum);
        int right_max = dfs(node->right, max_sum);

        // We use max(gain_from_left_subtree, 0) because we want to consider 
        // the gain only if it is positive. If it is negative, 
        // we ignore it or consider it as zero.
        left_max = max(left_max, 0);
        right_max = max(right_max, 0);

        max_sum = max(max_sum, node->val + left_max + right_max);

        return node->val + max(left_max, right_max);
    }
};